### ISSUE TITLE
Setting gamemode on player spawning from a plugin when player disconnected from game in spectator mode.

### ISSUE THREAD

**Issue by lkolbly** [cuberite/cuberite#3819]
Client version:   1.11.2
Server OS:        Ubuntu 17.04
Commit id:        Approx. e44597c9e46d65ebbb31ed7b6c8343bdef0a4515

### Steps to reproduce the behavior

Install this plugin:
~~~
function Initialize(Plugin)
   Plugin:SetName("Test")
   Plugin:SetVersion(1)

   cPluginManager.AddHook(cPluginManager.HOOK_PLAYER_SPAWNED, OnPlayerSpawn)

   return true
end

function OnPlayerSpawn(Player)
   Player:SetGameMode(gmSurvival)
end
~~~
(I called it "Test", but it doesn't matter)

This plugin essentially sets the game mode to survival when the player spawns, using the HOOK_PLAYER_SPAWNED callback.

Then, start the server and login.

Run "gamemode spectator <your name>" to set your gamemode to spectator.

Go up 5-ish blocks (though I don't think this part really matters)

Disconnect.

Reconnect.

### Expected behavior

On logging in, player becomes survival and drops to the ground, taking appropriate damage.

### Actual behavior

Player becomes survival (health & hunger bars appear) and falls through the ground into the void (taking damage along the way).

I suspect this is related to some sort of issue where the server sends the spectator capabilities (when the player logs in) and then immediately sends the survival capabilities (from the plugin hook), thus confusing the client in some way. (I'm looking at cClientHandle::Authenticate - maybe it would work if the player doesn't send capabilities (LoginSetGameMode) until after CallHookPlayerSpawned is called? I'll continue poking around)

### Workaround

One can apparently work around this by doing this instead:
~~~
TMP = -1

function Initialize(Plugin)
   Plugin:SetName("Test")
   Plugin:SetVersion(1)

   cPluginManager.AddHook(cPluginManager.HOOK_PLAYER_SPAWNED, OnPlayerSpawn)
   cPluginManager.AddHook(cPluginManager.HOOK_WORLD_TICK, OnWorldTick)

   return true
end

function OnPlayerSpawn(Player)
   --Player:SetGameMode(gmSurvival)
   TMP = 2
end

function OnWorldTick(World, TimeDelta)
   -- We only care about one world
   if World:GetDimension() ~= dimOverworld then
      return
   end

   --LOG(TMP)
   if TMP > 0 then
      TMP = TMP - 1
   elseif TMP == 0 then
      TMP = -1
      LOG("Setting to survival in world "..World:GetName())
      World:ForEachPlayer(function(Player)
	    LOG("Setting player survival");
	    Player:SetGameMode(gmSurvival)
      end)
   end
end
~~~
Essentially, just waiting for a couple of ticks before setting the player to survival.

**Comment by lkolbly** [comment]
Update: Using the player joined hook works as expected:
~~~
function Initialize(Plugin)
   Plugin:SetName("Test")
   Plugin:SetVersion(1)

   cPluginManager.AddHook(cPluginManager.HOOK_PLAYER_JOINED, OnPlayerJoin)

   return true
end

function OnPlayerJoin(Player)
   Player:SetGameMode(gmSurvival)
end
~~~

**Comment by williamhatcher** [comment]
So should this issue be closed then?

**Comment by bearbin** [comment]
No, there is still an issue. If it is on our side, it should be fixed, and if it is client-side the issue should be documented.

**Comment by tigerw** [comment]
I don't think so, but, \*inhales\*, _potentially_ an easy fix.
